src/nvim/tui/tui.c

```diff
@@ -0,0 +1,725 @@
+
+  if (nvim_override) {
+    str = get_term_option(ui, nvim_override);
+  } else if (unibi_index >= 0) {
+    if (unibi_index < unibi_string_begin_) {
+      str = unibi_get_ext_str(data->ut, (unsigned)unibi_index);
+    } else {
+      str = unibi_get_str(data->ut, (unsigned)unibi_index);
+    }
+  }
+
+  if (str) {
+    data->bufpos += unibi_run(str, data->params, data->buf + data->bufpos,
+        sizeof(data->buf) - data->bufpos);
+  }
+}
+
+static void out(UI *ui, const char *str)
+{
+  TUIData *data = ui->data;
+  data->bufpos += (size_t)snprintf(data->buf + data->bufpos,
+      sizeof(data->buf) - data->bufpos, "%s", str);
+}
+
+static void unibi_set_if_empty(unibi_term *ut, enum unibi_string str,
+    const char *val)
+{
+  if (!unibi_get_str(ut, str)) {
+    unibi_set_str(ut, str, val);
+  }
+}
+
+static void fix_terminfo(TUIData *data)
+{
+  unibi_term *ut = data->ut;
+
+  data->unibi_ext.enable_mouse = unibi_add_ext_str(ut, NULL,
+      "\x1b[?1002h\x1b[?1006h");
+  data->unibi_ext.disable_mouse = unibi_add_ext_str(ut, NULL,
+      "\x1b[?1002l\x1b[?1006l");
+
+  // Fill some empty slots with common terminal strings
+  unibi_set_if_empty(ut, unibi_cursor_address, "\x1b[%i%p1%d;%p2%dH");
+  unibi_set_if_empty(ut, unibi_exit_attribute_mode, "\x1b[0;10m");
+  unibi_set_if_empty(ut, unibi_set_a_foreground,
+      "\x1b[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m");
+  unibi_set_if_empty(ut, unibi_set_a_background,
+      "\x1b[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m");
+  unibi_set_if_empty(ut, unibi_enter_bold_mode, "\x1b[1m");
+  unibi_set_if_empty(ut, unibi_enter_underline_mode, "\x1b[4m");
+  unibi_set_if_empty(ut, unibi_enter_reverse_mode, "\x1b[7m");
+  unibi_set_if_empty(ut, unibi_bell, "\x07");
+  unibi_set_if_empty(data->ut, unibi_enter_ca_mode, "\x1b[?1049h");
+  unibi_set_if_empty(data->ut, unibi_exit_ca_mode, "\x1b[?1049l");
+  unibi_set_if_empty(ut, unibi_delete_line, "\x1b[M");
+  unibi_set_if_empty(ut, unibi_parm_delete_line, "\x1b[%p1%dM");
+  unibi_set_if_empty(ut, unibi_insert_line, "\x1b[L");
+  unibi_set_if_empty(ut, unibi_parm_insert_line, "\x1b[%p1%dL");
+  unibi_set_if_empty(ut, unibi_clear_screen, "\x1b[H\x1b[J");
+  unibi_set_if_empty(ut, unibi_clr_eol, "\x1b[K");
+  unibi_set_if_empty(ut, unibi_clr_eos, "\x1b[J");
+
+  char *term = getenv("TERM");
+  if (!term) {
+    return;
+  }
+
+#define STARTS_WITH(str, prefix) (!memcmp(str, prefix, sizeof(prefix) - 1))
+
+  if (STARTS_WITH(term, "xterm") || STARTS_WITH(term, "rxvt")) {
+    unibi_set_if_empty(ut, unibi_cursor_normal, "\x1b[?12l\x1b[?25h");
+    unibi_set_if_empty(ut, unibi_cursor_invisible, "\x1b[?25l");
+    unibi_set_if_empty(ut, unibi_flash_screen, "\x1b[?5h$<100/>\x1b[?5l");
+    unibi_set_if_empty(ut, unibi_exit_attribute_mode, "\x1b(B\x1b[m");
+    unibi_set_if_empty(ut, unibi_change_scroll_region, "\x1b[%i%p1%d;%p2%dr");
+    unibi_set_if_empty(ut, unibi_clear_screen, "\x1b[H\x1b[2J");
```

Comment:
I need to reorganize these lines so it will have preference over the more generic sequences